Translate API Reference of React.Component (#127)
reactjs/ja.reactjs.org: content/docs/reference-react-component.md

static getDerivedStateFromProps(props, state)
```

`getDerivedStateFromProps` は、初期マウント時とその後の更新時の両方で、`render()` メソッドを呼び出す直前に呼び出されます。state を更新するためにオブジェクトを返すか、何も更新しないために `null` を返すべきです。

このメソッドは、state が時間の経過とともに変化する props に依存するような[まれな使用例](/blog/2018/06/07/you-probably-dont-need-derived-state.html#when-to-use-derived-state)のために存在します。たとえば、以前と以降の子を比較してどちらの子をアニメーションするかを決定する `<Transition>` コンポーネントを実装するときに便利です。

state を派生させると冗長なコードにつながり、コンポーネントを考えるのが難しくなります。
[あなたがより簡単な方法に慣れていることを確認してください。](/blog/2018/06/07/you-probably-dont-need-derived-state.html)

* props の変更に応じて**副作用を実行する**必要がある場合は（データのフェッチやアニメーションなど）、代わりに [`componentDidUpdate`](#componentdidupdate) ライフサイクルを使用してください

* **プロパティが変更されたときにのみデータを再計算したい**場合は、[代わりにメモ化ヘルパーを使用してください](/blog/2018/06/07/you-probably-dont-need-derived-state.html#what-about-memoization)

> * **プロパティが変更されたときに何かの state を「リセット」したい**場合は、代わりにコンポーネントを[完全に制御する](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-controlled-component)か、または [`key` で全く制御しない](/blog/2018/06/07/you-probably-dont-need-derived-state.html#recommendation-fully-uncontrolled-component-with-a-key)かを検討してください

Author: ありがとうございます！「を使って〜」としないと意味が通りにくいと思います。

getSnapshotBeforeUpdate(prevProps, prevState)
```

`getSnapshotBeforeUpdate()` は、最後にレンダーされた出力が、たとえば DOM にコミットされる直前に呼び出されます。これはコンポーネントが変更される可能性があるときに、変更される前に DOM からある情報（たとえば、スクロール位置）を取得することを可能にします。このライフサイクルによって返された値はすべて、`componentDidUpdate()` へのパラメータとして渡されます。

このユースケースは一般的ではありませんが、スクロール位置を特別な方法で処理する必要があるチャットのスレッドのような UI で発生する可能性があります。

> 以下の例のように、スナップショット値（または `null`）が返されるべきです。

Author: まとめてしまいましたが、合っているべきだと思います。ありがとうございます。

`setState()` は、コンポーネントを更新するための即時のコマンドではなく、**要求**として考えてください。パフォーマンスをよくするために、React はそれを遅らせて、単一パスで複数のコンポーネントを更新することがあります。React は state の変更がすぐに適用されることを保証しません。

`setState()` は常にコンポーネントを直ちに更新するわけではありません。それはバッチ式に更新するか後で更新を延期するかもしれません。これは `setState()` を呼び出した直後に `this.state` を読み取ることが潜在的な危険になります。代わりに、`componentDidUpdate` または `setState` コールバック（`setState(updater, callback)`）を使用してください。どちらも更新が適用された後に起動することが保証されています。前の state に基づいて state を設定する必要がある場合は、下記の `updater` 引数についてお読みください。

`shouldComponentUpdate()` が `false` を返さない限り、`setState()` は常に再レンダーされます。ミュータブルなオブジェクトが使用されていて、条件付きレンダーロジックが `shouldComponentUpdate()` で実装できない場合、新しい state が前の state と異なるときにのみ `setState()` を呼び出すと、不要な再レンダーを回避できます。

最初の引数の `updater` 関数は次のようなシグネチャです。

```javascript
(state, props) => stateChange
```

`state` は、変更が適用されているときのコンポーネントの state への参照です。直接変更するべきではありません。代わりに、`state` と `props` からの入力に基づいて新しいオブジェクトを構築することによって変更を表現する必要があります。たとえば、`props.step` によって `state` の値を増加したいとします。

```javascript
this.setState((state, props) => {
  return {counter: state.counter + props.step};
});
```

updater 関数が受け取る `state` と `props` の両方が最新のものであることが保証されています。アップデータの出力は `state` と浅くマージされています。

> `setState()` の 2 番目のパラメータは、`setState` が完了してコンポーネントが再レンダリングされると実行されるオプションのコールバック関数です。通常、そのようなロジックには代わりに `componentDidUpdate()` を使用することをおすすめします。

Author: 「オプションの〜」とはあまり言わないかもしれません。ご指摘ありがとうございます！

component.forceUpdate(callback)
```

デフォルトでは、コンポーネントの state や props が変わると、コンポーネントは再レンダーされます。`render()` メソッドが他のデータに依存している場合は、`forceUpdate()` を呼び出してコンポーネントの再レンダーが必要であることを React に伝えることができます。

`forceUpdate()` を呼び出すと、`shouldComponentUpdate()` をスキップして、コンポーネントに対して `render()` が呼び出されます。これにより、それぞれの子の `shouldComponentUpdate()` メソッドを含む、子コンポーネントの通常のライフサイクルメソッドがトリガーされます。マークアップが変更された場合にのみ React は DOM を更新します。

> 通常は `forceUpdate()` の使用をすべて避け、`render()` の `this.props` と `this.state` からのみ読み取るようにしてください。

Reviewer: 言い回しを調整しました [IMO]

```suggestion
通常は全ての `forceUpdate()` の使用を避け、`render()` の `this.props` と `this.state` からのみ読み取るようにしてください。
```

Author: こちら、ご提案いただいた内容のほうが良いと思うので取り込ませていただきました。
また、「通常は `forceUpdate()` を使用せずに、」ではどうでしょうか？意訳しすぎでしょうか。

Reviewer: @lightnet328 
> また、「通常は `forceUpdate()` を使用せずに、」ではどうでしょうか？意訳しすぎでしょうか。

いいと思います！

};
```

`props.color` が提供されていない場合は、デフォルトで `'blue'` に設定されます。

```js
  render() {
    return <CustomButton /> ; // props.color は blue にセットされます
  }
```

`props.color` が null に設定されている場合、それは null のままになります。

```js
  render() {
    return <CustomButton color={null} /> ; // props.color は null のままになります
  }
```

* * *

### `displayName` {#displayname}

> `displayName` 文字列はデバッグメッセージに使用されます。通常、コンポーネントを定義する関数またはクラスの名前から推測されるため、明示的に設定する必要はありません。デバッグ目的で別の名前を表示する場合や、高次コンポーネントを作成する場合は、明示的に設定することをおすすめします。詳細については、[簡単なデバッグのために表示名をラップする](/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging)を参照してください。

Reviewer: 訳語の統一と言い回しの調整を行いました。

```suggestion
`displayName` 文字列はデバッグメッセージに使用されます。通常、コンポーネントを定義する関数またはクラスの名前から推測されるため、明示的に設定する必要はありません。デバッグ目的で別の名前を表示する場合や、高階コンポーネントを作成する場合には、明示的に設定したくなるかもしれません。詳細については、[簡単なデバッグのために表示名をラップする](/docs/higher-order-components.html#convention-wrap-the-display-name-for-easy-debugging)を参照してください。
```

* * *

## インスタンスプロパティ {#instance-properties-1}

### `props` {#props}

`this.props` には、このコンポーネントの呼び出し元によって定義された props が含まれています。props の紹介は [コンポーネントと props](/docs/components-and-props.html) を見てください。

> 特に、`this.props.children` は特別なプロパティで、通常はタグ自体ではなく JSX 式の子タグによって定義されます。

Reviewer: 言い回しの調整を行いました [IMO]

```suggestion
特に、`this.props.children` は特別なプロパティで、通常はタグ自体にではなく JSX 式の子タグによって定義されます。
```
